src/index.ts

```diff
@@ -635,33 +635,14 @@ class ThreadPool {
     });
 
     worker.on('error', (err : Error) => {
-      // Work around the bug in https://github.com/nodejs/node/pull/33394
-      worker.ref = () => {};
-
-      // In case of an uncaught exception: Call the callback that was passed to
-      // `postTask` with the error, or emit an 'error' event if there is none.
-      const taskInfos = [...workerInfo.taskInfos.values()];
-      workerInfo.taskInfos.clear();
-
-      // Remove the worker from the list and potentially start a new Worker to
-      // replace the current one.
-      this._removeWorker(workerInfo);
-
-      if (workerInfo.isReady() && !this.workerFailsDuringBootstrap) {
-        this._ensureMinimumWorkers();
-      } else {
-        // Do not start new workers over and over if they already fail during
-        // bootstrap, there's no point.
-        this.workerFailsDuringBootstrap = true;
-      }
+      this._onError(worker, workerInfo, err, false);
+    });
 
-      if (taskInfos.length > 0) {
-        for (const taskInfo of taskInfos) {
-          taskInfo.done(err, null);
-        }
-      } else {
-        this.publicInterface.emit('error', err);
-      }
+    worker.on('exit', (exitCode : number) => {
+      const err = new Error(`worker exited with code: ${exitCode}`);
```

Comment:
Maybe here we can distinguish based on the exit code.
If the user intends to gracefully shutdown its thread, it can do it by sending `0` as the exit code, otherwise when due to a failure, `1` will be returned, then it can be wrapped inside of an Error object, and Piscina can bootup a new thread

Reply:
Hmm, yeah I hadn't thought about a `0` exit code. That said, this code only errors out any existing unfinished tasks, as opposed to emitting an error over the `publicInterface` itself.

I'm not sure piscina should accept a `0` exit with unfinished tasks more gracefully than it already does in this PR - but maybe we'd rather silently retry the tasks once on a new thread, or something similar? I'm not familiar enough with piscina to have an opinion here; deferring to you.

Reply:
Hmm, no I think you're right, handling all of the errors and wrapping them for allowing piscina create a new thread sounds good 👍